### PR TITLE
[Feature] Support loading images and labels from lmdb file

### PR DESCRIPTION
**Pull request by Mountchicken**
## Motivation
As mentioned in #979 , reading images and tags from lmdb files is a very academia friendly feature and we now support this feature.

## Usage
- You can use existing lmdb datasets from [https://github.com/HCIILAB/Scene-Text-Recognition-Recommendations](https://github.com/HCIILAB/Scene-Text-Recognition-Recommendations), or you can convert the dataset stored in the folder to lmdb format using the [img2lmdb](#979) code we provide.
- There are two changes that need to be made in the config, I'll take the training of CRNN as an example:
- Firstly, in `conifgs/_base_/recog_datasets/`, your loader should use lmdb as `file_format`.
```python
lmdb_root = 'path to lmdb folder'
train = dict(
    type='OCRDataset',
    img_prefix=lmdb_root ,
    ann_file=lmdb_root ,
    loader=dict(
        type='AnnFileLoader',
        repeat=1,
        file_format='lmdb',
        parser=dict(
            type='LineJsonParser',
            keys=['filename', 'text'],
            keys_idx=[0, 1],
            separator=' ')),
    pipeline=None,
    test_mode=False)
```
- Then, In the pipeline file, `configs/_base_/recog_pipelines`, you should use `LoadImageFromLMDB` instead:
```python
train_pipeline = [
    dict(type='LoadImageFromLMDB', color_type='grayscale'),
    ...
```

  

## Tests

- [x] Tested the lmdb file generated from ocr_toy_dataset. The lmdb generation function is from #979 .
- [x] Tested a commonly used academic lmdb file CUTE80 from [https://github.com/HCIILAB/Scene-Text-Recognition-Recommendations](https://github.com/HCIILAB/Scene-Text-Recognition-Recommendations)

### CODE REVIEW

File: mmocr/datasets/utils/backend.py [hunk 1]
```diff
@@ -21,20 +21,35 @@ def __init__(self, lmdb_path, encoding='utf8'):
         self.lmdb_path = lmdb_path
         self.encoding = encoding
         env = self._get_env()
-        with env.begin(write=False) as txn:
-            self.total_number = int(
-                txn.get('total_number'.encode('utf-8')).decode(self.encoding))
+        try:
+            with env.begin(write=False) as txn:
```
[the remark this answers is not in the document]
Comment (Mountchicken) [reply mmocr/datasets/utils/backend.py:25]: good idea

File: mmocr/datasets/utils/backend.py [hunk 2]
```diff
@@ -21,20 +21,35 @@ def __init__(self, lmdb_path, encoding='utf8'):
...
-            line = txn.get(str(index).encode('utf-8')).decode(self.encoding)
-        return line
+            # this is for label only lmdb format
+            try:
+                line = txn.get(str(index).encode('utf-8')).decode(
+                    self.encoding)
+            # this is for image and label lmdb format
+            except Exception:
+                index = index + 1
+                label_key, filename = b'label-%09d' % index, r'%s' % index
```
[the remark this answers is not in the document]
Comment (Mountchicken) [reply mmocr/datasets/utils/backend.py:50]: Great idea

File: mmocr/datasets/pipelines/loading.py [hunk 3]
```diff
@@ -133,3 +134,57 @@ def __call__(self, results):
...
+
+    def __init__(self, color_type='color'):
+        self.color_type = color_type
+        self.env = None
+        self.txn = None
+
+    def __call__(self, results):
+        lmdb_index = results['img_info']['filename']
+        data_root = results['img_prefix']
+        img_key = 'image-%09d' % int(lmdb_index)
```
[the remark this answers is not in the document]
Comment (Mountchicken) [reply mmocr/datasets/pipelines/loading.py:155]: Great idea

File: mmocr/datasets/utils/backend.py [hunk 4]
```diff
@@ -40,15 +41,33 @@ def __getitem__(self, index):
...
-                index = index + 1
-                label_key, filename = b'label-%09d' % index, r'%s' % index
-                line = filename + ' ' + txn.get(label_key).decode()
+            # use initial_item to avoid error triggered frequently
+            if self.initial_item:
+                # this is for label only lmdb format
+                try:
+                    line = txn.get(str(index).encode('utf-8')).decode(
+                        self.encoding)
+                    self.label_only = True
```
Comment (gaotongxiao) [inline mmocr/datasets/utils/backend.py:50]: Can we move the assignment of this variable to __init__? We already know the format when retrieving `total_number` or `num-samples` from the dataset.

Comment (Mountchicken) [reply mmocr/datasets/utils/backend.py:50]: We can't tell if label_only from `total_numer` or `num-samples`. In `lmdb_converter`, whether it is label_only or not, I am using `total_number` to be uniform with the previous MMOCR's data.

**Review by gaotongxiao — COMMENTED** [review]
Need to update a brief guide in https://mmocr.readthedocs.io/en/latest/tutorials/dataset_types.html#ocrdataset

File: mmocr/datasets/utils/backend.py [hunk 5]
```diff
@@ -18,23 +19,78 @@ class LmdbAnnFileBackend:
...
-        return line
+            if self.deprecated_format:
+                line = txn.get(str(index).encode('utf-8')).decode(
+                    self.encoding)
+                filename, text = line.split(' ')
+                line = json.dumps({
+                    'filename': filename,
+                    'text': text
+                },
+                                  ensure_ascii=False)
```
Comment (gaotongxiao) [inline mmocr/datasets/utils/backend.py:79]: line break

**Comment by gaotongxiao** [comment]
@Mountchicken You can directly put your PR message to https://mmocr.readthedocs.io/en/latest/tutorials/dataset_types.html#ocrdataset as a usage example.

**Comment by Mountchicken** [comment]
> @Mountchicken You can directly put your PR message to https://mmocr.readthedocs.io/en/latest/tutorials/dataset_types.html#ocrdataset as a usage example.

Sure.

**Review by gaotongxiao — APPROVED** [review]

File: mmocr/datasets/utils/backend.py [hunk 6]
```diff
@@ -18,23 +19,76 @@ class LmdbAnnFileBackend:
     """
 
     def __init__(self, lmdb_path, encoding='utf8'):
+        """Currently we support two lmdb formats, one is the lmdb file with
+        only abels generated by txt2lmdb (deprecated), and one is the lmdb file
```
Comment (xinke-wang) [inline mmocr/datasets/utils/backend.py:23]: Typo
```suggestion
        only labels generated by txt2lmdb (deprecated), and one is the lmdb file
```

File: docs/en/tutorials/dataset_types.md [hunk 7]
```diff
@@ -2,16 +2,36 @@
...
 
-Here we show some examples of using different combination of `loader` and `parser`.
-
+Here we show an example to load lmdb annotations:
+```python
+lmdb_root = 'path to lmdb folder'
+train = dict(
+    type='OCRDataset',
+    img_prefix=lmdb_root ,
+    ann_file=lmdb_root ,
```
Comment (xinke-wang) [inline docs/en/tutorials/dataset_types.md:22]:
```suggestion
    img_prefix=lmdb_root,
    ann_file=lmdb_root,
```

**Review by xinke-wang — APPROVED** [review]

**Pull request merged**